### PR TITLE
PS-5364 feature: Update Hashicorp Vault plugin to use v2 API

### PR DESCRIPTION
**Pull request by percona-ysorokin**
https://jira.percona.com/browse/PS-5364

### CODE REVIEW

File: plugin/keyring_vault/tests/mtr/mount_point_service.inc [hunk 1]
```diff
@@ -1,69 +1,93 @@
...
-    system(qq#curl --http1.1 -L -H "X-Vault-Token: $token" $vault_ca_cert_opt --data '{"type":"generic"}' --request POST $vault_url/v1/sys/mounts/$secret_mount_point#);
+    my $postdata= "";
+    if ($imported_vault_server_mount_point_version eq "2")
+    {
+      $postdata= qq#{"type":"kv","options":{"version":"2"}}#;
+    } else {
+      $postdata= qq#{"type":"kv"}#;
+    }
+    system(qq#curl --http1.1 --insecure -L -H "X-Vault-Token: $imported_vault_admin_token" $vault_ca_cert_opt --data '$postdata' --request POST $imported_vault_conf_address/v1/sys/mounts/$vault_mount#) == 0
+      || die "Cannot create mount point";
```
Comment (oleksandr-kachan) [inline plugin/keyring_vault/tests/mtr/mount_point_service.inc]: I'd suggest using "or die ..." here instead. The "or" and "||" have different precedence in Perl and only "or" is usually used for flow control to avoid unexpected behavior.

File: plugin/keyring_vault/tests/mtr/mount_point_service.inc [hunk 2]
```diff
@@ -1,69 +1,93 @@
...
+    # kv to kv-v2 and therefore reduce the chances of
+    # 'Upgrading from non-versioned to versioned' Vault Server error
+    sleep(2) if $imported_vault_server_mount_point_version eq "2";
   }
-  elsif ($mount_point_service_op eq 'DELETE')
+  elsif ($imported_mount_point_service_op eq 'DELETE')
   {
-    system(qq#curl --http1.1 -L -H "X-Vault-Token: $token" $vault_ca_cert_opt -X DELETE $vault_url/v1/sys/mounts/$secret_mount_point#);
+    system(qq#curl --http1.1 --insecure -L -H "X-Vault-Token: $imported_vault_admin_token" $vault_ca_cert_opt -X DELETE $imported_vault_conf_address/v1/sys/mounts/$vault_mount#) == 0
+      || die "Cannot delete mount point";
```
Comment (oleksandr-kachan) [inline plugin/keyring_vault/tests/mtr/mount_point_service.inc]: Same here

File: plugin/keyring_vault/vault_credentials_parser.cc [hunk 3]
```diff
@@ -1,143 +1,314 @@
...
-      err_ss << "Could not parse credential file. Seems there is no value specified ";
-      err_ss << "for option " << option << " in line: " << line_number << '.';
+bool Vault_credentials_parser::parse(
+    const std::string &conf_file_path,
+    Vault_credentials &vault_credentials) const
+{
+  std::ifstream ifs(conf_file_path.c_str());
+  if (!ifs.is_open())
+  {
+    logger_->log(MY_ERROR_LEVEL, "Could not open credentials file.");
```
Comment (oleksandr-kachan) [inline plugin/keyring_vault/vault_credentials_parser.cc]: Should we also print file name here?

File: plugin/keyring_vault/vault_credentials_parser.h [hunk 4]
```diff
@@ -1,42 +1,27 @@
...
+ public:
+  Vault_credentials_parser(ILogger *logger) : logger_(logger) {}
+
+  bool parse(const std::string &conf_file_path,
+             Vault_credentials &vault_credentials) const;
+
+ private:
+  ILogger *logger_;
+
+  static const std::size_t max_file_size= 16384;
```
Comment (oleksandr-kachan) [inline plugin/keyring_vault/vault_credentials_parser.h:38]: Shall we keep such constants names in uppercase to easier distinguish them from regular variables?

File: plugin/keyring_vault/vault_curl.cc [hunk 5]
```diff
@@ -149,18 +92,78 @@ std::string Vault_curl::get_error_from_curl(CURLcode curl_code)
   return ss.str();
 }
 
+Secure_string Vault_curl::get_secret_url(const Secure_string &type_of_data) {
+  Secure_ostringstream oss_data;
+
+  DBUG_ASSERT(!vault_credentials_.get_vault_url().empty());
+  oss_data << vault_credentials_.get_vault_url() << "/v1/";
```
Comment (oleksandr-kachan) [inline plugin/keyring_vault/vault_curl.cc:101]: Just to recheck. Is "/v1/" here a mount point version? If so, shouldn't it be filled in based on secret_mount_point_version?

File: plugin/keyring_vault/vault_curl.cc [hunk 6]
```diff
@@ -194,71 +197,71 @@ bool Vault_curl::setup_curl_session(CURL *curl)
...
+  if (http_code == 404) {
+    *response = ""; // list returned empty list
     return false; 
   }
   *response = read_data_ss.str();
   return http_code / 100 != 2; // 2** are success return codes
 }
 
+bool Vault_curl::list_mount_points(Secure_string *response) {
+  return do_list(vault_credentials_.get_vault_url() + "/v1/sys/mounts",
```
Comment (oleksandr-kachan) [inline plugin/keyring_vault/vault_curl.cc]: Is it Ok to hardcode v1 in URL here?

File: plugin/keyring_vault/vault_parser_composer.cc [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+  doc.Parse(payload.c_str());
+  const rapidjson::Document &cdoc= doc;
+  if (cdoc.HasParseError())
+  {
+    logger->log(MY_ERROR_LEVEL, "Could not parse Vault Server response.");
+    return true;
+  }
+  if (!cdoc.IsObject())
+  {
+    logger->log(MY_ERROR_LEVEL, "Vault Server response iss not an Object");
```
Comment (oleksandr-kachan) [inline plugin/keyring_vault/vault_parser_composer.cc]: typo - "iss"

File: plugin/keyring_vault/vault_parser_composer.h [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */
+
+#ifndef MYSQL_VAULT_PARSER_H
```
Comment (oleksandr-kachan) [inline plugin/keyring_vault/vault_parser_composer.h]: It may be better to keep it consistent with file name - MYSQL_VAULT_PARSER_COMPOSER

File: unittest/gunit/keyring_vault/vault_environment.h [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    return mount_point_path_;
+  }
+
+  const std::string &get_admin_token() const { return admin_token_; }
+
+
+  static Vault_environment *create_instance()
+  {
+    if (instance_ != NULL)
+      return NULL;
```
Comment (oleksandr-kachan) [inline unittest/gunit/keyring_vault/vault_environment.h]: Shouldn't it just return already existing instance in this case?

File: unittest/gunit/keyring_vault/vault_environment.h [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+    instance_= new Vault_environment();
+    return instance_;
+  }
+  static Vault_environment *get_instance() { return instance_; }
```
Comment (oleksandr-kachan) [inline unittest/gunit/keyring_vault/vault_environment.h:67]: Can we just extend this method to create instance when needed and remove create_instance() at all?

File: sql/sys_vars.cc [hunk 11]
```diff
@@ -4686,7 +4686,7 @@ static Sys_var_version Sys_version(
 static char *server_version_suffix_ptr;
 static Sys_var_charptr Sys_version_suffix(
        "version_suffix", "version_suffix",
-       GLOBAL_VAR(server_version_suffix_ptr), NO_CMD_LINE,
+       GLOBAL_VAR(server_version_suffix_ptr), CMD_LINE(REQUIRED_ARG),
```
Comment (oleksandr-kachan) [inline sql/sys_vars.cc:4699]: What is this change needed for?

**Review by oleksandr-kachan — COMMENTED** [review]
Please take a look at some comments

File: plugin/keyring_vault/tests/mtr/mount_point_service.inc (hunk 1 above)
Comment (percona-ysorokin) [reply plugin/keyring_vault/tests/mtr/mount_point_service.inc]: fixed

File: plugin/keyring_vault/tests/mtr/mount_point_service.inc (hunk 2 above)
Comment (percona-ysorokin) [reply plugin/keyring_vault/tests/mtr/mount_point_service.inc]: fixed

File: plugin/keyring_vault/vault_curl.cc (hunk 5 above)
Comment (percona-ysorokin) [reply plugin/keyring_vault/vault_curl.cc:101]: No, `/v1/` is the version of Vault Server API. It is the same `v1` for both `kv` and `kv-v2` secret engines.

File: plugin/keyring_vault/vault_credentials_parser.h (hunk 4 above)
Comment (percona-ysorokin) [reply plugin/keyring_vault/vault_credentials_parser.h:38]: Personal preferences, STL/clib coding conventions.

File: plugin/keyring_vault/vault_credentials_parser.cc (hunk 3 above)
Comment (percona-ysorokin) [reply plugin/keyring_vault/vault_credentials_parser.cc]: added

File: plugin/keyring_vault/vault_curl.cc (hunk 6 above)
Comment (percona-ysorokin) [reply plugin/keyring_vault/vault_curl.cc]: Yes, `/v1/` is totally fine here - see previous comment. Moreover, mount point secret engine version auto-detection mechanism has been completely reworked in one of the later commits and `Vault_curl::list_mount_points()` method has been removed.

File: plugin/keyring_vault/vault_parser_composer.cc (hunk 7 above)
Comment (percona-ysorokin) [reply plugin/keyring_vault/vault_parser_composer.cc]: fixed

File: plugin/keyring_vault/vault_parser_composer.h (hunk 8 above)
Comment (percona-ysorokin) [reply plugin/keyring_vault/vault_parser_composer.h]: fixed

File: unittest/gunit/keyring_vault/vault_environment.h (hunk 10 above)
Comment (percona-ysorokin) [reply unittest/gunit/keyring_vault/vault_environment.h:67]: Not really, the problem is that once an instance of `Vault_environment` is passed to `::testing::AddGlobalTestEnvironment()`
```c++
::testing::AddGlobalTestEnvironment(Vault_environment::create_instance());
```
we are no longer responsible for its lifetime - Google Test environment is supposed to free this resource.
Therefore, this is not a canonical `singleton` pattern.

File: unittest/gunit/keyring_vault/vault_environment.h (hunk 9 above)
Comment (percona-ysorokin) [reply unittest/gunit/keyring_vault/vault_environment.h]: fixed

File: sql/sys_vars.cc (hunk 11 above)
Comment (percona-ysorokin) [reply sql/sys_vars.cc:4699]: This allows to create `keyring_vault` MTR test suite combinations (`plugin/keyring_vault/tests/mtr/combinations `)
```
[mount-v1]
version-suffix=mount-v1

[mount-v2]
version-suffix=mount-v2
...
```
by passing additional info to the server via not very meaningful variable `version-suffix`.
This variable `@@version_suffix` is parsed inside `plugin/keyring_vault/tests/mtr/parse_combination.inc`

**Comment by percona-ysorokin** [comment]
No GCA because of the missing pre-requisite commit.

**Comment by percona-ysorokin** [comment]
Jenkins build link
https://ps57.cd.percona.com/job/percona-server-5.7-param/405/
`main.mysqld--help-notwin` MTR test case fixed in the follow-up commit 6cc8b16

**Review by oleksandr-kachan — APPROVED** [review]
